app/controllers/social_previews_controller.rb

```diff
@@ -25,8 +26,7 @@ def user
   end
 
   def listing
-    @listing = ClassifiedListing.find(params[:id])
-    define_categories
+    @listing = ClassifiedListing.find(params[:id]).decorate
```

Comment:
Added a new decorator with 2 methods that will be used in the corresponding view.